### ISSUE TITLE
Displaying item access status in search results in DSpace 7

### ISSUE THREAD

**Issue by pilasou** [DSpace/dspace-angular#1566]
In DSpace 7, a user can only know if an item is open access or not by opening the item record. 

User must be able to know quickly in the UI if the main file of an item is Open Access or not. One first step would be to display the access status on search results.  

The item metadata dc.date available cannot be rely upon to determine access status as it is not consistently used by DSpace to display the Open Access date of the item (see https://groups.google.com/g/dspace-tech/c/k7qCSJVhRrU and https://github.com/DSpace/DSpace/issues/6583). 

In DSpace, access restrictions are set on bitstream. As an item can have one to many bitstream associated to it with different access restrictions, access status on the item should be based on the main file of the item (i.e. full text of an article, thesis, etc.). Hypothesis is that the first file is the main file in DSpace original bundle.  

4 access status can be displayed based on the restriction value used in DSPace 7 workflow:
<html>
<body>

Statut (anglais) | Description
-- | --
Open Access | Value Open access  in DSpace Workflow is selected for the file
Embargo | Value embargo in DSpace Workflow is selected for the file
Restricted | Value Administrator in DSpace Workflow is selected for the file
Metadata only | No bitstream associated

</body>
</html>

The following process can be used to determine the access status:  

1. Look at all READ policies of the primary/first bitstream, 
2. If one policy is bond to the Anonymous group and if the current date is between the start date and the end date of this policy, then item status is OpenAccess.  
3. Otherwise: if one policy for the Anonymous group has a start date inferior to the specific predefined date set in the configuration (for infinite embargo), then, it is an embargo item. 
4. If the start date is superior or equal the date specified in the configuration, then, the item is restricted. 
5. If no policy for the Anonymous group is set, the item is considered Restricted. 
6. Finally, if the item has no bitstream, the Item status will be Metadata only. 

A couple of parameters would be added to the configuration. First off, the access status display in search results can be activated or deactivated as a configuration parameter in DSpace. Next, the forever date would also be added to the configuration with a year, a month and a day. 

Access status will be displayed as tags as illustrated below:  
![access-status](https://user-images.githubusercontent.com/19291880/159006198-fabf5b64-40df-4fbf-a580-768668c71934.png)

**Comment by tdonohue** [comment]
@nibou230 : My immediate response is that you may hit major performance problems if you attempted to do most of the "logic" (behind these badges) in the Angular UI codebase.  I'd worry that, at best, you'd need to make several additional queries to the REST API _per Item_....which could be _very slow_ on any search/browse page that is displaying more than 10 items at once. (While the default is 10 items per page, it is configurable to display more.)

I suspect it'd have much better performance if the logic could be done on the backend.  We might consider indexing this info into Solr so that it can be returned as part of the Item response (e.g. whenever the Item is returned, it has an "accessStatus").  However, then the challenge would be in knowing when to trigger a re-index of the Item (i.e. how to know when the status has changed).  Unfortunately, embargoes could be the hardest to deal with, as when the embargo expires, there's no modification of the Item, so there's no way to know it needs a new "accessStatus" without checking if the embargo has expired yet.

We also need to keep in mind that some sites choose to add Embargoes or Access Restrictions to the *Item* instead of the Bitstream(s). So, the "Embargoed" and "Restricted" badges may be more complex than they appear, since the embargo/restriction could be at the Item level or the Bitstream level or potentially both...e.g. Item can be Restricted to a limited group of users, while it's Bitstream is also *Embargoed* until a date passes (Example use case: an Item may only be available "On Campus", but only after a specific date passes.  So, it's temporarily admin-only, but once the embargo passes, it's "On Campus" access.)

Overall, I like the general idea... but, I'm not sure how to easily implement it in DSpace 7.  It'd also need to be planned carefully to avoid performance problems.

Per @pilasou 's request via email, I'll add this to our DSpace 7 meeting agenda to see if others have ideas in that meeting.

**Comment by pilasou** [comment]
@tdonohue Thanks! Our goal is, as a starting point, to display the access status only for the anonymus group. This is the basis of the contribution we want to make. We want to implement it as a configuration in DSpace (people can easily switch on/off). We see some similarities with the Request Copy feature in Dspace (are we mistaken?) Let's discuss all this on Thurday then! :)

**Comment by nibou230** [comment]
@abollini Thanks! I'm happy you took the time to write this message.

It gives many hints at how we can implement this now and in the future for improvements. We will take the time to analyze all those possibilities considering the time we have left on the project.

This will help us a lot going forward!

**Comment by pnbecker** [comment]
Thanks to @tdonohue to make me aware of this issue here.

In DSpace 6 there was a feature that allowed to add visual indicators of different kinds to browse and search results. I know that it was used to mark items as Open Access depending on metadata like dc.rights or dc.rights.uri. As far as I know this is not ported to DSpace 7 yet. I just created an issue to discuess whether we want to port or to remove it from DSpace 7: https://github.com/DSpace/DSpace/issues/8244.

Maybe it is worth to take a look on the solution we had in DSpace 6 JSPUI for this (see the issue linked above for more details). If I remember correctly, the benefit of the solution in DSpace 6 was, that indicators for all kind of topics, not only for permissions and licenses, could be realized.

I'd be grateful if you could take a look and maybe add a comment to [DSpace/8244](https://github.com/DSpace/DSpace/issues/8244) whether you think the code might be useful for what you try to achieve here.

**Comment by pnbecker** [comment]
I don't know that JSPUI feature very well, but seeing class names like `org.dspace.app.itemmarking.ItemMarkingCollectionStrategy` and `org.dspace.app.itemmarking.ItemMarkingAvailabilityBitstreamStrategy` and taking a look on [item-markin.xml](https://github.com/DSpace/DSpace/blob/main/dspace/config/spring/api/item-marking.xml) suggest that it is easy to extend the ideas about what you want to use visual markers in search and browse results for and to make the feature discussed here a bit more general.

**Comment by nibou230** [comment]
@pnbecker Yes, I looked at the old code and this could be nice to port this feature in DSpace 7 in the future by converting all the badges to something more generic. Sadly, as you may know, I don't have much time and I won't be able to do and include this work in my future pull request. Maybe it could be discussed furthermore in the new issue.

**Comment by selensoftr** [comment]
is possible show Embargo Access Date in Bitstream File in Angular?

<img width="153" alt="EmbargoDate" src="https://github.com/DSpace/dspace-angular/assets/161182841/4f601699-1e0f-4436-9693-e3d4b0d22962">

**Comment by tdonohue** [comment]
@selensoftr : No, that's in another ticket https://github.com/DSpace/dspace-angular/issues/2413
